File: android/app/src/main/java/dev/vtlinh/ytkids/VideoStore.kt
package dev.vtlinh.ytkids

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase

/* The grid, on this device.

   This used to be a file of tab-separated lines per channel, next to the
   database. It is in the database now for one reason above the others: the
   Worker's reply is a delta, and answering "which videos do I already have"
   is a query rather than a file to re-read and re-parse. Rows and the channel
   they belong to also go together now — removing a channel removes its videos
   in the same place the approval lives.

   Not an optimisation, either way. A child opening the app on dropped wifi
   should still get the channels their parent approved rather than an empty
   screen they have no way to interpret or fix.

   The SQL lives in Schema.kt so it can be run against a real engine in a test;
   this file is only the Android plumbing around it. */
object VideoStore {

    private fun db(context: Context): SQLiteDatabase =
        ChannelStore.get(context).writableDatabase

    private val COLUMNS =
        arrayOf("video_id", "title", "published_at", "thumb_url", "channel_id")

    /* One channel's videos, in the order the Worker sent — which is upload
       order. The grid re-sorts by date across channels; the Channels tab shows
       one and this order is already right for it. */
    fun forChannel(context: Context, channelId: String): List<Video> = try {
        db(context).query(
            Schema.VIDEOS, COLUMNS, "channel_id = ?", arrayOf(channelId),
            null, null, "position ASC",
        ).use { c ->
            buildList { while (c.moveToNext()) add(c.toVideo()) }
        }
    } catch (e: Exception) {
        emptyList()
    }

    /* Everything, grouped, in ChannelStore's order — which is what the grid
       and the Channels tab are both built from. */
    fun byChannel(context: Context): Map<String, List<Video>> {
        val out = LinkedHashMap<String, List<Video>>()
        for (channel in ChannelStore.get(context).all()) {
            val videos = forChannel(context, channel.id)
            if (videos.isNotEmpty()) out[channel.id] = videos
        }
        return out
    }

    /* Just the ids, which is what goes to the Worker so it can leave out the
       details we already have. */
    fun knownIds(context: Context, channelId: String): List<String> = try {
        db(context).query(
            Schema.VIDEOS, arrayOf("video_id"), "channel_id = ?", arrayOf(channelId),
            null, null, "position ASC",
        ).use { c ->
            buildList { while (c.moveToNext()) add(c.getString(0)) }
        }
    } catch (e: Exception) {
        emptyList()
    }

    /* REPLACE the channel's list, rather than merge into it.
     *
     * The Worker's reply is the answer to "what does this channel have now",
     * so a video missing from it is a video the channel no longer lists — an
     * upload deleted, made private, or pushed past the hundred. Merging would
     * keep those forever and the grid would slowly fill with videos that no
     * longer exist, each one a tile that plays nothing.
     *
     * In one transaction: a delete that succeeded and an insert that did not
     * would empty a channel the child was looking at. */
    fun replace(context: Context, channelId: String, videos: List<Video>) {
        if (videos.isEmpty()) return
        val d = db(context)
        try {
            d.beginTransaction()
            d.delete(Schema.VIDEOS, "channel_id = ?", arrayOf(channelId))
            videos.forEachIndexed { position, v ->
                val values = ContentValues().apply {
                    put("video_id", v.id)
                    put("channel_id", channelId)
                    put("title", v.title)
                    v.publishedAt?.let { put("published_at", it) }
                    v.thumbUrl?.let { put("thumb_url", it) }
                    put("position", position)
                }
                /* CONFLICT_REPLACE on the id, because the same video appears in
                   two channels' lists after a collaboration and the primary key
                   is the video rather than the pair. Whichever channel was
                   written last owns the row; the grid collates by id anyway, so
                   the tile is the same either way. */
                d.insertWithOnConflict(
                    Schema.VIDEOS, null, values, SQLiteDatabase.CONFLICT_REPLACE,
                )
            }
            d.setTransactionSuccessful()
        } catch (e: Exception) {
            /* Leaves whatever was there. An empty grid is worse than a stale
               one. */
        } finally {
            try { d.endTransaction() } catch (e: Exception) {}
        }
    }

    /* A channel is no longer approved: its videos go with it, at once, rather
       than lingering until something else evicts them. */
    fun forget(context: Context, channelId: String) {
        try {
            db(context).delete(Schema.VIDEOS, "channel_id = ?", arrayOf(channelId))
        } catch (e: Exception) {}
    }

    private fun android.database.Cursor.toVideo() = Video(
        id = getString(0),
        title = getString(1),
        publishedAt = if (isNull(2)) null else getLong(2),
        thumbUrl = if (isNull(3)) null else getString(3),
    )
}

File: ios/TinyTube/VideoStore.swift
import Foundation
import TinyTubeCore

/* The grid, in SQLite. Counterpart of VideoStore.kt. */
enum VideoStore {

    private static let columns =
        "video_id, title, published_at, thumb_url"

    private static func video(_ r: Database.Row) -> Video {
        Video(
            id: r.string(0),
            title: r.string(1),
            publishedAt: r.intOrNil(2),
            thumbURL: r.stringOrNil(3)
        )
    }

    static func forChannel(_ channelId: String) -> [Video] {
        (try? Database.shared.read(
            """
            SELECT \(columns) FROM videos
            WHERE channel_id = ? ORDER BY position ASC
            """,
            [.text(channelId)],
            row: video
        )) ?? []
    }

    /* Every channel's list, keyed by channel. What `Library` collates into the
       grid — it wants them grouped, because the same video can be in two
       channels' lists after a collaboration. */
    static func byChannel() -> [String: [Video]] {
        let rows = (try? Database.shared.read(
            """
            SELECT channel_id, \(columns) FROM videos
            ORDER BY channel_id ASC, position ASC
            """,
            row: { r in
                (r.string(0), Video(
                    id: r.string(1),
                    title: r.string(2),
                    publishedAt: r.intOrNil(3),
                    thumbURL: r.stringOrNil(4)
                ))
            }
        )) ?? []

        var out: [String: [Video]] = [:]
        for (channelId, video) in rows { out[channelId, default: []].append(video) }
        return out
    }

    /* The ids this device already has for a channel. Sent to the Worker so the
       reply carries full details only for what is new — the difference between
       a 1.4 KB refresh and a 19 KB one. */
    static func knownIds(channelId: String) -> [String] {
        (try? Database.shared.read(
            "SELECT video_id FROM videos WHERE channel_id = ? ORDER BY position ASC",
            [.text(channelId)],
            row: { $0.string(0) }
        )) ?? []
    }

    /* The Worker's reply REPLACES a channel's videos rather than merging into
       them. It is the answer to what the channel has NOW, so a video missing
       from it was deleted, made private, or pushed past the hundred; merging
       would slowly fill the grid with tiles that play nothing.
     *
     * An empty list changes nothing. A reply that parsed to nothing is a
     * failure wearing the shape of an answer, and a stale grid beats an empty
     * one — which is also what a phone with no signal gets. */
    static func replace(channelId: String, videos: [Video]) {
        guard !videos.isEmpty else { return }

        /* What this channel's posters were BEFORE the reply, so the ones that
           do not survive it can be deleted from the phone.
         *
         * ImageStore never expires — a URL fetched once is kept for good — so
         * removal is the only thing that reclaims anything. A channel that
         * posts daily pushes a video off the end of its hundred every day, and
         * without this each one leaves a poster on disk for the life of the
         * install. Read here rather than after the write: afterwards there is
         * nothing left to read.
         *
         * BOTH SIDES USE `thumbnailURL`, NOT `thumb_url`. The column is
         * nullable — Uploads.thumb answers nil for a reply with no thumb, or one
         * from an unexpected host — and the grid then draws the DERIVED
         * i.ytimg.com URL, which is the name the JPEG is cached under. Reading
         * the column alone made those posters invisible to this comparison, so a
         * video dropping off the end of the hundred left its file on disk for
         * the life of the install. */
        let before: [String] = (try? Database.shared.read(
            "SELECT video_id, thumb_url FROM videos WHERE channel_id = ?",
            [.text(channelId)],
            row: { Video(id: $0.string(0), title: "", thumbURL: $0.stringOrNil(1)).thumbnailURL }
        )) ?? []

        do {
            try Database.shared.transaction {
                try Database.shared.write(
                    "DELETE FROM videos WHERE channel_id = ?", [.text(channelId)]
                )
                for (position, v) in videos.enumerated() {
                    /* REPLACE on the id, because the same video appears in two
                       channels' lists after a collaboration and the primary key
                       is the video rather than the pair. Whichever channel was
                       written last owns the row; the grid collates by id
                       anyway, so the tile is the same either way. */
                    try Database.shared.write(
                        """
                        INSERT OR REPLACE INTO videos
                            (video_id, channel_id, title, published_at, thumb_url, position)
                        VALUES (?, ?, ?, ?, ?, ?)
                        """,
                        [
                            .text(v.id),
                            .text(channelId),
                            .text(v.title),
                            .int(v.publishedAt),
                            .text(v.thumbURL),
                            .int(Int64(position)),
                        ]
                    )
                }
            }
        } catch {
            /* Leaves whatever was there. An empty grid is worse than a stale
               one, so a failed replace is not allowed to become a deletion.
               Nothing is pruned either — dropping the pictures for a write that
               rolled back would blank tiles the grid is still showing. */
            return
        }

        let kept = Set(videos.map { $0.thumbnailURL })
        ImageStore.forget(before.filter { !kept.contains($0) })
    }

    /* A channel is no longer approved: its videos go with it, at once, rather
       than lingering until something else evicts them. */
    static func forget(channelId: String) {
        try? Database.shared.write(
            "DELETE FROM videos WHERE channel_id = ?", [.text(channelId)]
        )
    }
}
